public/app/features/live/live.ts

```diff
@@ -0,0 +1,113 @@
+import Centrifuge, { PublicationContext } from 'centrifuge/dist/centrifuge.protobuf';
+import SockJS from 'sockjs-client';
+import { GrafanaLiveSrv, setGrafanaLiveSrv, ChannelHandler } from '@grafana/runtime';
+import { PartialObserver, Unsubscribable, Subject, BehaviorSubject } from 'rxjs';
+import { KeyValue } from '@grafana/data';
+
+interface Channel<T = any> {
+  subject: Subject<T>;
+  subscription?: Centrifuge.Subscription;
+}
+
+class CentrifugeSrv implements GrafanaLiveSrv {
+  centrifuge: Centrifuge;
+  channels: KeyValue<Channel> = {};
+  connectionListeners: BehaviorSubject<boolean>;
+
+  constructor() {
+    console.log('connecting....');
+    // TODO: better pick this from the URL
+    this.centrifuge = new Centrifuge(`http://${location.host}/live/sockjs`, {
+      debug: true,
+      sockjs: SockJS,
+    });
+    this.centrifuge.setToken('ABCD');
+    this.centrifuge.connect(); // do connection
+    this.connectionListeners = new BehaviorSubject<boolean>(this.centrifuge.isConnected());
+
+    // Register global listeners
+    this.centrifuge.on('connect', this.onConnect);
+    this.centrifuge.on('disconnect', this.onDisconnect);
+    this.centrifuge.on('publish', this.onServerSideMessage);
+  }
+
+  //----------------------------------------------------------
+  // Internal functions
+  //----------------------------------------------------------
+
+  onConnect = (context: any) => {
+    console.log('CONNECT', context);
+    this.connectionListeners.next(true);
+  };
+
+  onDisconnect = (context: any) => {
+    console.log('onDisconnect', context);
+    this.connectionListeners.next(false);
+  };
+
+  onServerSideMessage = (context: any) => {
+    console.log('Publication from server-side channel', context);
+  };
+
+  //----------------------------------------------------------
+  // Exported functions
+  //----------------------------------------------------------
+
+  /**
+   * Is the server currently connected
+   */
+  isConnected = (): boolean => {
```

Comment:
would make all functions here live on the prototype instead (ie not define them using equal signs & arrow function), a bit allergic to this function definition style, would only use it as an exception for functions passed to other components

Reply:
"a bit allergic"  :rofl:   HA -- yes, the functions as properties voodo is awkward.  

I updated the things that feel like methods to be real methods, but kept the functions that get passed as callbacks as properties -- IIUC, the property syntax helps with making sure `this` refers to the right scope without needing to bind